### PR TITLE
Add multiserial agent functionality

### PR DESCRIPTION
**Pull request by Acuadros95**

### CODE REVIEW

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/426/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/427/)

File: include/uxr/agent/transport/custom/CustomAgent.hpp [hunk 1]
```diff
@@ -69,20 +70,23 @@ class CustomAgent : public Server<CustomEndPoint>
...
      * @param message_length Number of bytes to be sent.
+     * @param serial_fd Serial file descriptor.
      * @param transport_rc Transport return code, to be filled by the user.
      * @return ssize_t Number of sent bytes.
      */
     using SendMsgFunction = std::function<ssize_t (
         const CustomEndPoint* /*destination_endpoint*/,
         uint8_t* /*buffer*/,
         size_t /*message_length*/,
+        uint8_t /*write file descriptor*/,
```
Comment (pablogs9) [inline include/uxr/agent/transport/custom/CustomAgent.hpp:89]:
```suggestion
        uint8_t /*serial_fd*/,
```

File: include/uxr/agent/transport/custom/CustomAgent.hpp [hunk 2]
```diff
@@ -69,20 +70,23 @@ class CustomAgent : public Server<CustomEndPoint>
         uint8_t* /*buffer*/,
         size_t /*buffer_length*/,
         int /*timeout*/,
+        uint8_t /*read file descriptor*/,
```
Comment (pablogs9) [inline include/uxr/agent/transport/custom/CustomAgent.hpp:73]:
```suggestion
        uint8_t /*serial_fd*/,
```

File: include/uxr/agent/transport/stream_framing/StreamFramingProtocol.hpp [hunk 3]
```diff
@@ -92,30 +92,35 @@ class FramingIO
...
      * @param buffer Raw octet buffer to write into.
      * @param message_length Number of bytes to be written.
+     * @param serial_fd Serial file descriptor.
      * @param transport_rc Return code to be set by the callback logic.
      * @return size_t Number of written bytes.
      */
     using WriteCallback = std::function<ssize_t (
             uint8_t* /*buffer*/,
             size_t /*message_length*/,
+            uint8_t /*write file descriptor*/,
```
Comment (pablogs9) [inline include/uxr/agent/transport/stream_framing/StreamFramingProtocol.hpp:102]:
```suggestion
            uint8_t /*serial_fd*/,
```

File: include/uxr/agent/transport/stream_framing/StreamFramingProtocol.hpp [hunk 4]
```diff
@@ -92,30 +92,35 @@ class FramingIO
...
      * @param timeout Read timeout.
+     * @param serial_fd Serial file descriptor.
      * @param transport_rc Return code to be set by the callback logic.
      * @return size_t Number of read bytes.
      */
     using ReadCallback = std::function<ssize_t (
             uint8_t* /*buffer*/,
             size_t /*buffer_length*/,
             int /*timeout*/,
+            uint8_t /*read file descriptor*/,
```
Comment (pablogs9) [inline include/uxr/agent/transport/stream_framing/StreamFramingProtocol.hpp:118]:
```suggestion
            uint8_t /*serial_fd*/,
```

**Review by pablogs9 — CHANGES_REQUESTED** [review]

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/427/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/428/)

File: src/cpp/transport/serial/MultiSerialAgentLinux.cpp [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                if (0 < bytes_read)
+                {
+                    input_packet.message.reset(new InputMessage(buffer_, static_cast<size_t>(bytes_read)));
+                    input_packet.source = MultiSerialEndPoint(it->first, remote_addr);
+                    rv = true;
+
+                    uint32_t raw_client_key;
+                    if (Server<MultiSerialEndPoint>::get_client_key(input_packet.source, raw_client_key))
+                    {
+                        UXR_AGENT_LOG_MESSAGE(
```
Comment (pablogs9) [inline src/cpp/transport/serial/MultiSerialAgentLinux.cpp:104]: Maybe here a reference to the serial descriptor.

File: src/cpp/transport/serial/MultiSerialAgentLinux.cpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+    if ((0 < bytes_written) && (
+         static_cast<size_t>(bytes_written) == output_packet.message->get_len()))
+    {
+        rv = true;
+
+        uint32_t raw_client_key;
+        if (Server<MultiSerialEndPoint>::get_client_key(output_packet.destination, raw_client_key))
+        {
+            UXR_AGENT_LOG_MESSAGE(
```
Comment (pablogs9) [inline src/cpp/transport/serial/MultiSerialAgentLinux.cpp:153]: Maybe here a reference to the serial descriptor?

File: src/cpp/transport/serial/MultiSerialAgentLinux.cpp [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+ssize_t MultiSerialAgent::write_data(
+        uint8_t serial_fd,
+        uint8_t* buf,
+        size_t len,
+        TransportRc& transport_rc)
+{
+    size_t rv = 0;
+    ssize_t bytes_written = ::write(serial_fd, buf, len);
+    if ((0 < bytes_written)  && size_t(bytes_written) == len)
```
Comment (pablogs9) [inline src/cpp/transport/serial/MultiSerialAgentLinux.cpp:210]:
```suggestion
    if ((0 < bytes_written) && size_t(bytes_written) == len)
```

Comment (pablogs9) [inline src/cpp/transport/serial/MultiSerialAgentLinux.cpp:210]: Check `FramingIO::transport_write``I think that this function can send less bytes than `len`

**Review by pablogs9 — CHANGES_REQUESTED** [review]

File: src/cpp/transport/serial/MultiSerialAgentLinux.cpp (hunk 7 above)
Comment (Acuadros95) [reply src/cpp/transport/serial/MultiSerialAgentLinux.cpp]: True (write_buffer_pos_ is len parameter):
https://github.com/eProsima/Micro-XRCE-DDS-Agent/blob/ed66bb0ffecbb1f234fefb21ee9af4a3e2c7cb86/src/cpp/transport/stream_framing/StreamFramingProtocol.cpp#L397-L402

Modified on Serial and Multiserial modes: cd18d34

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/428/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/429/)

File: src/cpp/transport/serial/MultiSerialAgentLinux.cpp (hunk 5 above)
Comment (Acuadros95) [reply src/cpp/transport/serial/MultiSerialAgentLinux.cpp]: Updated 46a62d3:
```
[1621849003.217650] debug    | MultiSerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x4A705497, fd: 13, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 4A 02 00 00 80
[1621849003.218211] debug    | MultiSerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x50A5B13C, fd: 3, len: 20, data: 
0000: 81 80 49 02 07 01 08 00 02 53 00 05 45 02 00 00 00 00 08 40
[1621849003.218308] debug    | MultiSerialAgentLinux.cpp | send_message             | [** <<SER>> **]        | client_key: 0x50A5B13C, fd: 3, len: 13, data: 
0000: 81 00 00 00 0A 01 05 00 11 00 00 00 80
[1621849003.218478] debug    | MultiSerialAgentLinux.cpp | recv_message             | [==>> SER <<==]        | client_key: 0x4A705497, fd: 13, len: 16, data: 
0000: 81 80 4A 02 07 01 08 00 02 54 00 05 46 02 00 00
```

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/429/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/430/)

File: src/cpp/transport/serial/MultiSerialAgentLinux.cpp [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+    struct timeval timeout_ = { 1, 0 }; /* Seconds, Microseconds */
+    bool rv = false;
+
+    utils::SharedLock lk(framing_mtx);
+    fd_set fds = read_fds;
+    int ret = select(framing_io.rbegin()->first+1, &fds, NULL, NULL, &timeout_);
+    
+    if (0 < ret)
+    {
+        for (std::map<int, FramingIO>::iterator it = framing_io.begin(); it != framing_io.end(); it++)
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply src/cpp/transport/serial/MultiSerialAgentLinux.cpp]: 1bdfe87

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/430/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/431/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/431/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/432/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/433/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/434/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/434/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/435/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/437/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/438/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/438/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/439/)

File: include/uxr/agent/transport/Server.hpp [hunk 9]
```diff
@@ -77,6 +77,16 @@ class Server : public Agent, public SessionManager<EndPoint>
             int timeout,
             TransportRc& transport_rc) = 0;
 
+    virtual bool recv_message(
+            std::vector<InputPacket<EndPoint>>& input_packet,
+            int timeout,
+            TransportRc& transport_rc) {
+                    (void) input_packet;
+                    (void) timeout;
+                    (void) transport_rc;
```
Comment (pablogs9) [inline include/uxr/agent/transport/Server.hpp:86]:
```suggestion
            std::vector<InputPacket<EndPoint>>& /* input_packet */,
            int /* timeout */,
            TransportRc& /* transport_rc */) {
```

File: include/uxr/agent/transport/serial/MultiSerialAgentLinux.hpp [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+#endif
+
+    bool recv_message(
+            InputPacket<MultiSerialEndPoint>& input_packet,
+            int timeout,
+            TransportRc& transport_rc) final { 
+                    (void) input_packet;
+                    (void) timeout;
+                    (void) transport_rc;
+                    return false;
```
Comment (Acuadros95) [inline include/uxr/agent/transport/serial/MultiSerialAgentLinux.hpp:69]:
```suggestion
    bool recv_message(
            InputPacket<MultiSerialEndPoint>& /* input_packet */,
            int /* timeout */,
            TransportRc& /* transport_rc */) final { 
                    return false; 
```

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Linux/441/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Agent%20Manual%20Windows/442/)

**Review by pablogs9 — APPROVED** [review]

**Review by pablogs9 — APPROVED** [review]

**Pull request merged**